Mathlib/Data/Finset/Antidiagonal.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir and María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández, Bhavik Mehta
+-/
+
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finsupp.Defs
+import Mathlib.Data.Finsupp.Interval
+import Mathlib.Algebra.Order.Sub.Defs
+
+/-! # Antidiagonal with values in general types
+
+Let `n : μ`, where `μ` is a canonically ordered add monoid with locally finite order.
+(For example, one may take `μ` to be `ℕ`, more generally `σ →₀ ℕ`.)
+
+* For `s : Finset ι`, we define `Finset.piAntidiagonal s n` as the `Finset (ι → μ)`
+of functions with support in `s` whose sum is equal to `n`.
+
+* We define `Finset.antidiagonal n : Finset (μ × μ)` of pairs adding to `n`.
+
+These definitions generalize `Finset.Nat.antidiagonal` and `Finsupp.antidiagonal`
+which are defined in `Mathlib.Data.Finset.NatAntidiagonal`
+and `Mathlib.Data.Finsupp.Antidiagonal` and we make the comparisons there.
+
+This definition does not exactly match with that of `Multiset.antidiagonal` which is
+defined in `Mathlib.Data.Multiset.Antidiagonal`, because of the multiplicities.
+Indeed, by counting multiplicities, `Multiset α` is equivalent to `α →₀ ℕ`,
+but `Finset.antidiagonal` and `Multiset.antidiagonal` will return different objects.
+For example, for `s : Multiset ℕ := {0,0,0}`, `Multiset.antidiagonal s` has 8 elements
+but `Finset.antidiagonal s` has only 4.
+
+```lean
+def s : Multiset ℕ := {0, 0, 0}
+#eval (Finset.antidiagonal s).card -- 4
+#eval Multiset.card (Multiset.antidiagonal s) -- 8
+```
+
+This is mostly taken from a file of Bhavik Mehta,
+Archive/Wiedijk100Theorems/Partition.lean,
+where `piAntidiagonal` was called `cut`
+-/
+
+open scoped BigOperators
+
+open Finset Function
+
+/-- The class of monoids with an antidiagonal -/
+class HasMulAntidiagonal (μ : Type*) [Monoid μ] where
+  /-- The antidiagonal function -/
+  (antidiagonal : μ → Finset (μ × μ))
+  /-- A pair belongs to `antidiagonal n` iff the product of its components is equal to `n` -/
+  (mem_antidiagonal : ∀ (n : μ) (a : μ × μ), a ∈ antidiagonal n ↔ a.fst * a.snd = n)
+
+/-- The class of additive monoids with an antidiagonal -/
+class HasAntidiagonal (μ : Type*) [AddMonoid μ] where
+  /-- The antidiagonal function -/
+  (antidiagonal : μ → Finset (μ × μ))
+  /-- A pair belongs to `antidiagonal n` iff the sum of its components is equal to `n` -/
+  (mem_antidiagonal : ∀ (n : μ) (a : μ × μ), a ∈ antidiagonal n ↔ a.fst + a.snd = n)
+
+variable {μ : Type _}
+  [CanonicallyOrderedAddMonoid μ]
+  [LocallyFiniteOrder μ] [DecidableEq μ]
+
+variable {ι : Type _} [DecidableEq ι] [DecidableEq (ι → μ)]
+
+namespace Finset
+
+/-- In a canonically ordered add monoid, the finset of pairs of elements of `μ` with sum `n` -/
+def antidiagonal (n : μ) : Finset (μ × μ) :=
+  Finset.filter (fun uv => uv.fst + uv.snd = n) (Finset.product (Iic n) (Iic n))
```

Comment:
Here's how this becomes restated using the new typeclasses:
```suggestion
/-- In a canonically ordered add monoid, the antidiagonal can be constructed by filtering.

Note that this is not an instance, as for some times a more efficient algorithm is available. -/
abbrev antidiagonalOfLocallyFinite : HasAntidiagonal μ :=
  antidiagonal n := Finset.filter (fun uv => uv.fst + uv.snd = n) (Finset.product (Iic n) (Iic n))
  mem_antidiagonal n := sorry
```

Reply:
at end of line 73, is it `:=` or `where`?

Reply:
that required a lot of prefixes `antidiagonalOfLocallyFinite` later on, but that worked.